# Data Cleaning and Preparation

## Australian Energy Market Analytics

This notebook prepares electricity price and demand data from the Australian
Energy Market Operator (AEMO) for exploratory and statistical analysis.

### Objectives

- Load and inspect the raw AEMO price and demand data
- Assess data quality and identify missing or duplicate observations
- Convert timestamps into an appropriate datetime format
- Validate electricity demand and spot price values
- Create time-based features for subsequent analysis
- Export a cleaned dataset for exploratory analysis

## 1. Import Libraries

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## 2. Load the Dataset

The dataset contains five-minute electricity demand and Regional Reference
Price (RRP) observations from the Australian Energy Market Operator (AEMO).

In [3]:
file_path = "../data/raw/PRICE_AND_DEMAND_202608_NSW1.csv"

df = pd.read_csv(file_path)

df.head()

,REGION,SETTLEMENTDATE,TOTALDEMAND,RRP,PERIODTYPE
0,NSW1,2026/08/01 00:05:00,8898.91,89.88,TRADE
1,NSW1,2026/08/01 00:10:00,9032.33,95.05,TRADE
2,NSW1,2026/08/01 00:15:00,8989.12,88.88,TRADE
3,NSW1,2026/08/01 00:20:00,9036.72,88.88,TRADE
4,NSW1,2026/08/01 00:25:00,8946.67,84.79,TRADE


## 3. Initial Data Inspection
### 3.1. Dataset Dimensions

In [4]:
df.shape

print(f"Number of observations: {df.shape[0]:,}")
print(f"Number of variables: {df.shape[1]}")

Number of observations: 7,776
Number of variables: 5


The raw dataset contains **7,776** observations across **5** variables.
Each observation represents a five-minute electricity market interval.

### 3.2. Inspect column types

In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7776 entries, 0 to 7775
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   REGION          7776 non-null   str    
 1   SETTLEMENTDATE  7776 non-null   str    
 2   TOTALDEMAND     7776 non-null   float64
 3   RRP             7776 non-null   float64
 4   PERIODTYPE      7776 non-null   str    
dtypes: float64(2), str(3)
memory usage: 516.5 KB


`SETTLEMENTDATE` is currently stored as an object rather than a datetime
variable. This will be converted before performing time-series analysis.

In [7]:
df["SETTLEMENTDATE"] = pd.to_datetime(df["SETTLEMENTDATE"])
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7776 entries, 0 to 7775
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   REGION          7776 non-null   str           
 1   SETTLEMENTDATE  7776 non-null   datetime64[us]
 2   TOTALDEMAND     7776 non-null   float64       
 3   RRP             7776 non-null   float64       
 4   PERIODTYPE      7776 non-null   str           
dtypes: datetime64[us](1), float64(2), str(2)
memory usage: 372.2 KB


### Variable Description

| Variable | Description |
|---|---|
| `REGION` | National Electricity Market region identifier |
| `SETTLEMENTDATE` | Timestamp for the five-minute market interval |
| `TOTALDEMAND` | Total regional electricity demand in MW |
| `RRP` | Regional Reference Price in $/MWh |
| `PERIODTYPE` | Type of market interval |

## 4. Data Quality Assessment

### 4.1 Missing Values

In [8]:
missing_values = pd.DataFrame({
    "Missing Values": df.isnull().sum(),
    "Missing Percentage": df.isnull().mean() * 100
})

missing_values

,Missing Values,Missing Percentage
REGION,0,0.0
SETTLEMENTDATE,0,0.0
TOTALDEMAND,0,0.0
RRP,0,0.0
PERIODTYPE,0,0.0


No missing values were identified in the raw dataset, so no imputation or
row removal is required at this stage.

### 4.2 Duplicate Observations

In [9]:
duplicate_count = df.duplicated().sum()

print(f"Number of duplicate rows: {duplicate_count}")

Number of duplicate rows: 0


## 5. Export Cleaned Dataset

The cleaned dataset is exported for use in subsequent exploratory
and statistical analyses.

In [10]:
output_path = "../data/processed/clean_price_demand.csv"

df.to_csv(output_path, index=False)

print(f"Cleaned dataset saved to: {output_path}")

Cleaned dataset saved to: ../data/processed/clean_price_demand.csv
